packages/headless/src/controllers/commerce/context/headless-context.ts

```diff
@@ -0,0 +1,187 @@
+  trackingId: nonEmptyString,
+  language: nonEmptyString,
+  currency: nonEmptyString,
+  clientId: nonEmptyString,
+  user: new RecordValue({
+    values: {
+      userId: nonEmptyString,
+      email: nonEmptyString,
+      userIp: nonEmptyString,
+      userAgent: nonEmptyString,
+    },
+  }),
+  view: new RecordValue({
+    options: { required: true },
+    values: {
+      url: requiredNonEmptyString,
+    }
+  })
+});
+
+export interface ContextOptions {
+  trackingId?: string;
+  language?: string;
+  currency?: string;
+  clientId?: string;
+  user?: User;
+  view: View;
+}
+
+export interface User {
+  userId?: string;
+  email?: string;
+  userIp?: string;
+  userAgent?: string;
+}
+
+export interface View {
+  url: string;
+}
+
+export interface ContextProps {
+  /**
+   * The initial options that should be applied to this `Context` controller.
+   */
+  options: ContextOptions;
+}
+
+/**
+ * The `Context` controller allows the end user to configure context data.
+ */
+export interface Context extends Controller {
+  /**
+   * Sets the tracking id.
+   * @param trackingId - The new tracking id.
+   */
+  setTrackingId(trackingId: string): void;
+
+  /**
+   * Sets the language.
+   * @param language - The new language.
+   */
+  setLanguage(language: string): void;
+
+  /**
+   * Sets the currency.
+   * @param currency - The new currency.
+   */
+  setCurrency(currency: string): void;
+
+  /**
+   * Sets the client id.
+   * @param clientId - The new client id.
+   */
+  setClientId(clientId: string): void;
+
+  /**
+   * Sets the user.
+   * @param user - The new user.
+   */
+  setUser(user: User): void;
+
+  /**
+   * Sets the view.
+   * @param view - The new view.
+   */
+  setView(view: View): void;
+
+  /**
+   * A scoped and simplified part of the headless state that is relevant to the `Context` controller.
+   */
+  state: ContextState;
+}
+
+export interface ContextState {
+  trackingId?: string;
+  language?: string;
+  currency?: string;
+  clientId?: string;
+  user?: User;
+  view: View;
+}
+
+export type ContextControllerState = Context['state'];
+
+/**
+ * Creates a `ContextController` controller instance.
+ *
+ * @param engine - The headless commerce engine.
+ * @param props - The configurable `ContextController` properties.
+ * @returns A `ContextController` controller instance.
+ */
+export function buildContext(
+  engine: CommerceEngine,
+  props: ContextProps
+): Context {
+  if (!loadBaseContextReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  const getState = () => engine.state;
+
+  const options = {
+    ...props.options,
+  };
+
+  validateOptions(engine, optionsSchema, options, 'buildContext');
+
+  dispatch(
+    setContext({
+      ...options,
+    })
+  );
+
+  return {
+    ...controller,
+
+    get state() {
+      const {trackingId, language, currency, clientId, user, view} = getState().commerceContext;
```

Comment:
Are there properties we would not want to be returned here from the context ?

Or that we would like to modify/do some processing here before returning them ?

If not, could we simply do:

```suggestion
      get state() {
          return getState().commerceContext;
      }
```